src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs

```diff
@@ -135,93 +135,174 @@ private void LoadTestsFromAnExtension(string extensionAssembly, IEnumerable<stri
+            }
+            catch (Exception e)
+            {
+                var message = string.Format(
+                    CultureInfo.CurrentUICulture,
+                    CrossPlatEngineResources.ExceptionFromLoadTests,
+                    discovererType.Name,
+                    e.Message);
+
+                logger.SendMessage(TestMessageLevel.Error, message);
+                EqtTrace.Error("DiscovererEnumerator.DiscoverTestsFromSingleDiscoverer: {0} ", e);
+            }
+        }
+
+        private static bool TryToLoadDiscoverer(LazyExtension<ITestDiscoverer, ITestDiscovererCapabilities> discoverer, IMessageLogger logger, out Type discovererType)
+        {
+            discovererType = null;
+
+            // See if discoverer can be instantiated successfully else move next.
+            try
+            {
+                discovererType = discoverer.Value.GetType();
+            }
+            catch (Exception e)
+            {
+                var mesage = string.Format(
+                    CultureInfo.CurrentUICulture,
+                    CrossPlatEngineResources.DiscovererInstantiationException,
+                    e.Message);
+                logger.SendMessage(TestMessageLevel.Warning, mesage);
+                EqtTrace.Error("DiscovererEnumerator.LoadTestsFromAnExtension: {0} ", e);
+
+                return false;
+            }
+
+            return true;
+        }
+
+        private void RecordTotalTestsDiscoveredByCurrentAdapter(long currentTotalTests, LazyExtension<ITestDiscoverer, ITestDiscovererCapabilities> discoverer,
+            ref bool discoverersFromDeprecatedLocations, ref double totalAdaptersUsed)
+        {
+            if (this.discoveryResultCache.TotalDiscoveredTests > currentTotalTests)
+            {
+                var totalDiscoveredTests = this.discoveryResultCache.TotalDiscoveredTests - currentTotalTests;
+                this.requestData.MetricsCollection.Add(
+                    string.Format("{0}.{1}", TelemetryDataConstants.TotalTestsByAdapter,
+                        discoverer.Metadata.DefaultExecutorUri), totalDiscoveredTests);
+                if (!CrossPlatEngine.Constants.DefaultAdapters.Contains(discoverer.Metadata.DefaultExecutorUri.ToString(),
+                    StringComparer.OrdinalIgnoreCase))
                 {
-                    var message = string.Format(
-                        CultureInfo.CurrentUICulture,
-                        CrossPlatEngineResources.ExceptionFromLoadTests,
-                        discovererType.Name,
-                        e.Message);
+                    var discovererLocation = discoverer.Value.GetType().GetTypeInfo().Assembly.GetAssemblyLocation();
 
-                    logger.SendMessage(TestMessageLevel.Error, message);
-                    EqtTrace.Error("DiscovererEnumerator.LoadTestsFromAnExtension: {0} ", e);
+                    discoverersFromDeprecatedLocations |= Path.GetDirectoryName(discovererLocation)
```

Comment:
If discoverersFromDeprecatedLocations is true, then no need to do anything like getting discovererLocation etc (as we are doing OR in discoverersFromDeprecatedLocations bool)

Reply:
Good catch. Fixed it.